src/datastructures/nalgebra.rs

```diff
@@ -8,6 +9,27 @@ use super::{AdjacencyMatrix, Edge};
 #[derive(Debug, PartialEq)]
 pub struct NAMatrix(pub DMatrix<f64>);
 
+impl NAMatrix {
+    /// Create from set of x/y points
+    pub fn from_points(points: &[[f64; 2]]) -> Self {
+        let mut matrix = NAMatrix::from_dim(points.len());
+        for i in 0..points.len() {
+            let from = &points[i];
+            let weights: Vec<f64> = points
+                .par_iter()
+                .skip(i + 1)
+                .map(|to| ((from[0] - to[0]).powi(2) + (from[1] - to[1]).powi(2)).sqrt())
+                .collect();
+            weights.iter().enumerate().for_each(|(j, weight)| {
+                let j = i + j + 1;
+                matrix[(i, j)] = *weight;
+                matrix[(j, i)] = *weight;
+            });
+        }
```

Comment:
```suggestion
        points.iter().enumerate().for_each(|(i, from)| {
            points.iter().skip(i + 1).enumerate().for_each(|(j, to)| {
                let j = i + j + 1;
                let weight = ((from[0] - to[0]).powi(2) + (from[1] - to[1]).powi(2)).sqrt();
                // uses the fact, that the euclidean distance is symmetric
                matrix[(i, j)] = weight;
                matrix[(j, i)] = weight;
            });
        });
```
Thanks for your suggestion. I've noticed several things in your code:
1. you are allocating a vector for every iteration of the loop
1. you are using much direct array indexing
1. you are using the parallel iterator only on the inner loop

I suggest the following things:
1. remove the vector allocation, streamline the whole iterator pipeline.
1. transform direct array indexing to the use of iterators
1. remove the parallel iterator, create a separate function for that.